source/create.js

```diff
@@ -1,64 +1,61 @@
 'use strict';
-const extend = require('extend');
-const is = require('@sindresorhus/is');
+const URLGlobal = typeof URL === 'undefined' ? require('url').URL : URL; // TODO: Use the `URL` global when targeting Node.js 10
+const errors = require('./errors');
+const assignOptions = require('./assign-options');
 const asStream = require('./as-stream');
 const asPromise = require('./as-promise');
-const errors = require('./errors');
 const normalizeArguments = require('./normalize-arguments');
 
-const assignOptions = (defaults, options = {}) => {
-	const opts = extend(true, {}, defaults, options);
+const next = (path, options) => {
+	let url = path;
 
-	if (Reflect.has(options, 'headers')) {
-		for (const [key, value] of Object.entries(options.headers)) {
-			if (is.nullOrUndefined(value)) {
-				delete opts.headers[key];
-				continue;
-			}
-		}
+	if (options.baseUrl) {
+		url = new URLGlobal(path, options.baseUrl);
+	}
+
+	options = normalizeArguments(url, options);
+
+	if (options.stream) {
+		return asStream(options);
 	}
 
-	return opts;
+	return asPromise(options);
 };
 
-const create = (defaults = {}) => {
+const create = defaults => {
+	if (!defaults.handler) {
+		defaults.handler = next;
+	}
+
 	function got(url, options) {
 		try {
-			options = assignOptions(defaults, options);
-			const normalizedArgs = normalizeArguments(url, options);
-
-			if (normalizedArgs.stream) {
-				return asStream(normalizedArgs);
-			}
-
-			return asPromise(normalizedArgs);
+			options = assignOptions(defaults.options, options);
+			return defaults.handler(url, options, next);
 		} catch (error) {
 			return Promise.reject(error);
 		}
 	}
 
-	got.create = (options = {}) => create(assignOptions(defaults, options));
+	got.create = create;
+	got.extend = (options = {}) => create({
+		options: assignOptions(defaults.options, options),
+		methods: defaults.methods,
+		handler: defaults.handler
+	});
 
 	got.stream = (url, options) => {
-		options = assignOptions(defaults, options);
-		return asStream(normalizeArguments(url, options));
+		options = assignOptions(defaults.options, options);
+		options.stream = true;
+		return defaults.handler(url, options, next);
 	};
 
-	const methods = [
-		'get',
-		'post',
-		'put',
-		'patch',
-		'head',
-		'delete'
-	];
-
-	for (const method of methods) {
+	for (const method of defaults.methods) {
 		got[method] = (url, options) => got(url, {...options, method});
 		got.stream[method] = (url, options) => got.stream(url, {...options, method});
 	}
 
 	Object.assign(got, errors);
+	Object.assign(got, {defaults});
```

Comment:
Exposing this and having it used in the closure could lead to some unexpected behavior if defaults are modified: 

```js
test('tampering with defaults', async t => {
	const instance = got.create({
		handler: got.defaults.handler,
		methods: got.defaults.methods,
		options: { ...got.defaults.options,
			baseUrl: 'example'
		}
	});

	const instance2 = instance.create({
		handler: instance.defaults.handler,
		methods: instance.defaults.methods,
		options: instance.defaults.options
	});

	// Tamper Time
	instance.defaults.options.baseUrl = 'http://google.com'

	t.is(instance.defaults.options.baseUrl, instance2.defaults.options.baseUrl);
});
```

I think I'd expect two instances created with `.create` to be independent and not impact each other. Similar tampering could be performed on the global `got.defaults` that the module exports.

Thoughts?

Reply:
You're right. We should deep freeze those or clone.

Reply:
We can deep clone it with the `extend` dependency, I think.

Reply:
Deep cloning may lead to false information because it's still writable. Used deep freeze.